### PR TITLE
[Annotation plugin] Implement MapStyleObserverPlugin to listen style load event to reload layer and source

### PR DESCRIPTION
**Pull request by Chaoba**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).
Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>[Annotation plugin] Implement MapStyleObserverPlugin to listen style load event to reload layer and source</changelog>`.

### Summary of changes
This pr implement MapStyleObserverPlugin listen in annotation plugin to prevent reload failed issue.

### CODE REVIEW

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 1]
```diff
@@ -162,7 +165,7 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
     }
   }
 
-  protected fun initLayerAndSource() {
+  @Synchronized protected fun initLayerAndSource() {
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:168]: Synchronized not needed, all those events will happen on main thread.
https://github.com/mapbox/mapbox-maps-android/blob/da2ea096429959c2159a955042d5bfc222763922/sdk/src/main/java/com/mapbox/maps/NativeMapObserver.kt#L31-L33

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:168]: In the last commit, we got the following error on firebase, it seems we have sync issue in MapChange and getStyle callback. Maybe we could remove one to remove this `@Synchronized`, need to rethink it.

`java.lang.RuntimeException: Add source failed: Source mapbox-android-circle-source-1 already exists
	at com.mapbox.maps.extension.style.sources.Source.bindTo(Source.kt:59)
	at com.mapbox.maps.extension.style.sources.SourceKt.addSource(Source.kt:185)
	at com.mapbox.maps.plugin.annotation.AnnotationManagerImpl.initLayerAndSource(AnnotationManagerImpl.kt:177)
	at com.mapbox.maps.plugin.annotation.AnnotationManagerImpl$1$1.invoke(AnnotationManagerImpl.kt:114)
	at com.mapbox.maps.plugin.annotation.AnnotationManagerImpl$1$1.invoke(AnnotationManagerImpl.kt:42)
	at com.mapbox.maps.plugin.MapDelegateProviderImpl$getStyle$1.onStyleLoaded(MapDelegateProviderImpl.kt:20)
	at com.mapbox.maps.MapboxMap.getStyle(MapboxMap.kt:226)`

Comment (kiryldz) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:168]: Would be great to reproduce the error and find the root cause but for me it seems that `synchronized` is not needed here (if looking at code) and may just find the root cause for that error even harder to detect in future.

Comment (Chaoba) [reply plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt]: Yeah, find a new solution to remove `synchronized`, and works well.

File: plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt [hunk 2]
```diff
@@ -185,6 +181,19 @@ abstract class AnnotationManagerImpl<G : Geometry, T : Annotation<G>, S : Annota
     }
 
     updateSource()
+
+    // Listen to map change event to reload layer and source in the first init
+    if (!mapChangeListenerAdded) {
+      mapChangeListenerAdded = true
+      delegateProvider.mapListenerDelegate.addOnMapChangedListener {
```
Comment (kiryldz) [inline plugin-annotation/src/main/java/com/mapbox/maps/plugin/annotation/AnnotationManagerImpl.kt:188]: Should this listener be cleared as others in `onDestroy`? If yes - then flag should be reset in `onDestroy` as well.

**Comment by pengdev** [comment]
@Chaoba Regarding the approach, instead of adding a new map event listener, I think we could add `MapStyleObserverPlugin` interface to the AnnotationPlugin, and propagate the style changed event to the underlying annotation managers to re-add the layers/sources.  This delegate is also used to update the UI in LocationComponentPlugin.

**Review by pengdev — APPROVED** [review]
LGTM, please test locally as well 👍

**Review by kiryldz — APPROVED** [review]

**Pull request merged**